### PR TITLE
 undeploy endpoint with tests SRX-M5GOA4

### PR DESCRIPTION
**Pull request by medo-freiheit**
Added un-deploy endpoint to completely remove a service.

There are two checks before un-deploy an application:
1) check locks ( ignore env locks )
2) check that all releases are un-deployed

### CODE REVIEW

File: services/cd-service/pkg/repository/transformer.go [hunk 1]
```diff
@@ -234,6 +238,62 @@ func (c *CreateUndeployApplicationVersion) Transform(fs billy.Filesystem) (strin
...
+
+	configs, err := (&State{Filesystem: fs}).GetEnvironmentConfigs()
+
+	for env := range configs {
+		appDir := environmentApplicationDirectory(fs, env, u.Application)
+		locksDir := fs.Join(appDir, "locks")
+		undeployFile := fs.Join(appDir, "version", "undeploy")
+
+		if entries, _ := fs.ReadDir(locksDir); entries != nil {
+			return "", fmt.Errorf("UndeployApplication: error cannot un-deploy application '%v' unlock the application lock in the '%v' release first", u.Application, env)
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/transformer.go:272]:
```suggestion
			return "", fmt.Errorf("UndeployApplication: error cannot un-deploy application '%v' unlock the application lock in the '%v' environment first", u.Application, env)
```

File: services/cd-service/pkg/repository/transformer_test.go [hunk 2]
```diff
@@ -39,6 +39,163 @@ const (
...
+				&UndeployApplication{
+					Application: "app1",
+				},
+			},
+			expectedError:     "",
+			expectedCommitMsg: "application 'app1' was deleted successfully",
+			shouldSucceed:     true,
+		},
+		{
+			Name: "Undeploy application where there is an application locks shouldn't work",
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/transformer_test.go:82]:
```suggestion
			Name: "Undeploy application where there is an application lock, shouldn't work",
```

File: services/cd-service/pkg/repository/transformer_test.go [hunk 3]
```diff
@@ -39,6 +39,163 @@ const (
 	additionalVersions = 7
 )
 
+func TestUndeployApplicationErrors(t *testing.T) {
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/transformer_test.go:42]: I really like the way you tested this!

**Review by sven-urbanski-freiheit-com — CHANGES_REQUESTED** [review]
.

File: services/cd-service/pkg/repository/transformer_test.go [hunk 4]
```diff
@@ -39,6 +39,163 @@ const (
...
+					},
+				},
+				&CreateEnvironmentLock{
+					Environment: "dev",
+					LockId:      "22133",
+					Message:     "test",
+				},
+				&CreateUndeployApplicationVersion{
+					Application: "app1",
+				},
```
[the remark this answers is not in the document]
Comment (medo-freiheit) [reply services/cd-service/pkg/repository/transformer_test.go]: Yes! we are still able to add a lock even it's an un-deploy version, Thanks!

File: services/cd-service/pkg/repository/transformer_test.go (hunk 3 above)
Comment (medo-freiheit) [reply services/cd-service/pkg/repository/transformer_test.go:42]: I liked it too 🚀

**Review by sven-urbanski-freiheit-com — APPROVED** [review]

**Comment by sven-urbanski-freiheit-com** [comment]
Looks good! Please update the commit message before merging. It should be about "what does the endpoint do", e.g something like this:

Added undeploy endpoint to completely remove a service

This endpoint only works when there are no app locks... and ...

**Pull request merged**